libuavcan/include/libuavcan/time.hpp

```diff
@@ -14,153 +15,252 @@
     }
 
-    static DurationType fromMicrosecond(std::int64_t us)
+    static Type fromMicrosecond(USecT us)
     {
-        DurationType d;
+        Type d;
         d.usec_ = us;
         return d;
     }
 
-    static DurationType fromMillisecond(std::int64_t ms)
+    USecT toMicrosecond() const
     {
-        return fromMicrosecond(ms * 1000);
+        return usec_;
     }
 
-    std::int64_t toMicrosecond() const
+    USecT toMillisecond() const
     {
-        return usec_;
+        return usec_ / static_cast<USecT>(1000);
+    }
+
+    Type getAbs() const
+    {
+        return Type::fromMicrosecond(std::abs(usec_));
+    }
+
+    bool operator==(const Type& r) const
+    {
+        return usec_ == r.usec_;
+    }
+
+    bool operator!=(const Type& r) const
+    {
+        return !operator==(r);
     }
 
-    std::int64_t toMillisecond() const
+    bool operator<(const Type& r) const
     {
-        return usec_ / 1000;
+        return usec_ < r.usec_;
     }
 
-    DurationType getAbs() const
+    bool operator>(const Type& r) const
     {
-        return DurationType::fromMicrosecond((usec_ < 0) ? (-usec_) : usec_);
+        return usec_ > r.usec_;
     }
 
-    bool isPositive() const
+    bool operator<=(const Type& r) const
     {
-        return usec_ > 0;
+        return usec_ <= r.usec_;
     }
 
-    bool isNegative() const
+    bool operator>=(const Type& r) const
     {
-        return usec_ < 0;
+        return usec_ >= r.usec_;
     }
 
-    bool isZero() const
+    Type operator+(const Type& r) const
     {
-        return usec_ == 0;
+        return fromMicrosecond(util::saturating_add(usec_, r.usec_));
     }
 
-    bool operator==(const DurationType& r) const
+    Type operator-(const Type& r) const
+    {
+        return fromMicrosecond(util::saturating_sub(usec_, r.usec_));
+    }
+
+    Type operator-() const
+    {
+        return fromMicrosecond(-usec_);
+    }
+
+    Type& operator+=(const Type& r)
+    {
+        *this = *this + r;
+        return *static_cast<Type*>(this);
+    }
+
+    Type& operator-=(const Type& r)
+    {
+        *this = *this - r;
+        return *static_cast<Type*>(this);
+    }
+};
+
+class LIBUAVCAN_EXPORT Monotonic : public Base<Monotonic>
+{};
+
+}  // namespace duration
+
+namespace time
+{
+/**
+ * Protected base class for time values.
+ *
+ * @tparam Type          The type of the derived class. This must be an empty type.
+ *                       All storage will be provided by this base class.
+ * @tparam DType         The type of duration used for this time type. Time is concrete and duration
+ *                       is relative.
+ * @tparam USecT         The datatype returned when retrieving time from
+ *                       realizations of this base class. This type must be unsigned.
+ */
+template <typename Type, typename DType, typename USecT = uint64_t>
+class Base
+{
+    USecT usec_;
+
+protected:
+    ~Base() {}
+
+    Base()
+        : usec_(0)
```

Comment:
```suggestion
```
HIC++ 12.4.2, C++CG C.45

Reply:
Actually, by C.21 and because we want to define the default ctor as protected I need to implement all [default operations](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#cctor-constructors-assignments-and-destructors)

Reply:
But the constructor can be reduced to `Base() = default;`, can it not? There is no point putting static asserts there. My underlying concern is that the initializer list is used where it could be avoided. Or perhaps I am overreacting and this issue is so minor it shouldn't even be discussed?

Reply:
Actually, my compiler refuses to allow the static_assert outside of the ctor.

```
libuavcan/include/libuavcan/time.hpp:34:19: error: invalid application of 'sizeof' to an incomplete type 'libuavcan::duration::Monotonic'
    static_assert(sizeof(Type) == sizeof(USecT),
```

Reply:
Okay, it makes sense. What do we do with the field initializers? Not just here but in general in similar cases.

Reply:
I would seem, according to C.21, that we must provide them all if we provide any. The compilers will complain about missing or out of order initializers. We should also get Coverity scans running again which are really good at catching this sort of thing.